src/utils.rs

```diff
@@ -151,15 +150,27 @@ pub(crate) fn positive_bn_random_from_transcript(
     }
 }
 
-/// Generate a random BigNumber in the range 1..N-1 (Z_N^*) (non-zero)
-pub(crate) fn random_bn_in_z_star<R: RngCore + CryptoRng>(rng: &mut R, n: &BigNumber) -> BigNumber {
-    for _ in 0..CRYPTOGRAPHIC_RETRY_MAX {
-        let bn = BigNumber::from_rng(n, rng);
-        if bn != BigNumber::zero() {
-            return bn;
-        }
-    }
-    BigNumber::zero()
+/// Generate a random `BigNumber` that is in the multiplicative group of integers modulo `n`.
+///
+/// This method draws a random element in the range `[0, n-1]`, then checks that it is
+/// non-zero and coprime with `n`.
+/// Coprimality is tested by checking that the least common multiple is equal to the product of
+/// `n` and the element.
+///
+/// Note: In this application, `n` is typically the product of two primes. If the drawn element
+/// is coprime with `n`, then the caller has accidentally stumbled upon the factorization of `n`!
+/// This is a security issue when `n` is someone else's Paillier modulus, but it is highly
```

Comment:
This doesn't quite say what you mean it to say.
- If the drawn element is _not_ coprime with `n` and the element is not `0 mod n`, then you have accidentally stumbled upon the factorization, the chance of which is basically 0.